drivers/sensor/lsm6dsr/lsm6dsr.h

```diff
@@ -0,0 +1,112 @@
+/* ST Microelectronics LSM6DSO 6-axis IMU sensor driver
+ *
+ * Copyright (c) 2019 STMicroelectronics
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Datasheet:
+ * https://www.st.com/resource/en/datasheet/lsm6dsr.pdf
+ */
+
+#ifndef ZEPHYR_DRIVERS_SENSOR_LSM6DSR_LSM6DSR_H_
+#define ZEPHYR_DRIVERS_SENSOR_LSM6DSR_LSM6DSR_H_
+
+#include <zephyr/drivers/sensor.h>
+#include <zephyr/types.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/spi.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/util.h>
+#include <stmemsc.h>
+#include "lsm6dsr_reg.h"
+
+#if DT_ANY_INST_ON_BUS_STATUS_OKAY(spi)
+#include <zephyr/drivers/spi.h>
+#endif /* DT_ANY_INST_ON_BUS_STATUS_OKAY(spi) */
+
+#if DT_ANY_INST_ON_BUS_STATUS_OKAY(i2c)
+#include <zephyr/drivers/i2c.h>
+#endif /* DT_ANY_INST_ON_BUS_STATUS_OKAY(i2c) */
+
+#define LSM6DSR_EN_BIT					0x01
+#define LSM6DSR_DIS_BIT					0x00
+
+/* Accel sensor sensitivity grain is 61 ug/LSB */
+#define SENSI_GRAIN_XL				(61LL)
+
+/* Gyro sensor sensitivity grain is 4375 udps/LSB */
+#define SENSI_GRAIN_G				(4375LL)
+#define SENSOR_PI_DOUBLE			(SENSOR_PI / 1000000.0)
+#define SENSOR_DEG2RAD_DOUBLE		(SENSOR_PI_DOUBLE / 180)
+#define SENSOR_G_DOUBLE				(SENSOR_G / 1000000.0)
+
+struct lsm6dsr_config {
+	stmdev_ctx_t ctx;
+	union {
+#if DT_ANY_INST_ON_BUS_STATUS_OKAY(i2c)
+		const struct i2c_dt_spec i2c;
+#endif
+#if DT_ANY_INST_ON_BUS_STATUS_OKAY(spi)
+		const struct spi_dt_spec spi;
+#endif
+	} stmemsc_cfg;
+	uint8_t accel_pm;
+	uint8_t accel_odr;
+	uint8_t accel_range;
+	uint8_t gyro_pm;
+	uint8_t gyro_odr;
+	uint8_t gyro_range;
+	uint8_t drdy_pulsed;
+#ifdef CONFIG_LSM6DSR_TRIGGER
+	const struct gpio_dt_spec gpio_drdy;
+	uint8_t int_pin;
+	bool trig_enabled;
+#endif /* CONFIG_LSM6DSR_TRIGGER */
+};
+
+struct lsm6dsr_data {
+	const struct device *dev;
+	int16_t acc[3];
+	uint32_t acc_gain;
+	uint8_t accel_pm;
+	int16_t gyro[3];
+	uint32_t gyro_gain;
+#if defined(CONFIG_LSM6DSR_ENABLE_TEMP)
+	int16_t temp_sample;
+#endif
+
+	float accel_freq;
+	uint8_t accel_fs;
+	float gyro_freq;
+	uint8_t gyro_fs;
+
+#ifdef CONFIG_LSM6DSR_TRIGGER
+	struct gpio_callback gpio_cb;
+	sensor_trigger_handler_t handler_drdy_acc;
```

Comment:
It might not be necessary to store any of the `sensor_trigger_handler_t` inside of data. It's a function pointer, so it will never change, and it isn't used anywhere outside of `sensor_trigger_set()` from what I can tell. You may be able to reduce the trigger to a single pointer, if an array of them is declared in `LSM6DSR_DEFINE()`

Reply:
It's set by user when registering a callback.  `lsm6dsr_handle_interrupt()` calls it.

Reply:
Ah yes - I see that now. Somehow missed that before.